Wikipedia/UI-V5/WMFArticleHeaderImageGalleryViewController.m

```diff
@@ -95,45 +98,141 @@ - (UICollectionViewCell*)collectionView:(UICollectionView*)collectionView
-    .then(^id (UIImage* image) {
+    .then(^id (WMFImageDownload* download) {
+        @strongify(self);
+        UIImage* image = download.image;
         if (!self) {
             return [NSError cancelledError];
+        }
+        BOOL shouldAnimate = ![download.origin isEqualToString:[WMFImageDownload imageOriginMemory]];
+        if (!imageMetadata.didDetectFaces) {
+            DDLogVerbose(@"Running face detection for %@", imageMetadata.sourceURL);
+            @weakify(self);
+            return [self.faceDetector wmf_detectFeaturelessFacesInImage:image].then(^(NSArray* faces) {
+                @strongify(self);
+                [imageMetadata setNormalizedFaceBoundsFromFeatures:faces inImage:image];
+                [imageMetadata save];
+                NSParameterAssert(imageMetadata.didDetectFaces);
+                [self setImage:image
+                 inCellAtIndexPath:indexPath
+                   centeringBounds:imageMetadata.firstFaceBounds
+                          animated:shouldAnimate];
+            });
         } else {
-            if (!imageMetadata.allNormalizedFaceBounds) {
-                return [self.faceDetector wmf_detectFeaturelessFacesInImage:image].then(^(NSArray* faces) {
-                    @strongify(self);
-                    [imageMetadata setNormalizedFaceBoundsFromFeatures:faces inImage:image];
-                    [imageMetadata save];
-                    [self setImage:image centeringBounds:imageMetadata.firstFaceBounds forCellAtIndexPath:indexPath];
-                });
-            } else {
-                [self setImage:image centeringBounds:imageMetadata.firstFaceBounds forCellAtIndexPath:indexPath];
-                return nil;
-            }
+            DDLogVerbose(@"Setting image %@ after retrieving from %@", imageMetadata.sourceURL, download.origin);
+            [self setImage:image
+             inCellAtIndexPath:indexPath
+               centeringBounds:imageMetadata.firstFaceBounds
+                      animated:shouldAnimate];
+            return nil;
         }
+    })
+    .catch(^(NSError* error) {
+        // TODO: show error in UI
+        DDLogError(@"Failed to fetch image from %@. %@", [imageMetadata sourceURL], error);
     });
-    return cell;
 }
 
-- (void)setImage:(UIImage*)image centeringBounds:(CGRect)normalizedCenterBounds forCellAtIndexPath:(NSIndexPath*)path {
-    WMFImageCollectionViewCell* cell = (WMFImageCollectionViewCell*)[self.collectionView cellForItemAtIndexPath:path];
+/**
+ *  Attempt to populate the given `cell` with the image for `metadata` if it's stored in memory.
+ *
+ *  @param cell      The cell to populate.
+ *  @param indexPath The indexPath where the cell resides.
+ *  @param metadata  The metadata which specifies the image URL and face detection information.
+ *
+ *  @return `YES` if the image was set from memory, `NO` if a fetch or other processing needs to take place.
+ */
+- (BOOL)setCachedImageForCell:(WMFImageCollectionViewCell*)cell
```

Comment:
Similar to: https://github.com/wikimedia/wikipedia-ios/pull/69/files#r37236899 

At first glance, it looks like you are running enough logic here that would necessitate adding another method - like checking if face detection ran.

But when I looked where this method is actually used, it seems that doing the returning `NO` when `didDetectFaces=NO` on line 176 here is causing us to conflate face detection with fetching the image: https://github.com/wikimedia/wikipedia-ios/pull/69/files#r37237100

So if that is removed, we are back to really just 2 method calls in here. 

All of this is to say, that it just seems you can rework 
`- (void)     setImage:(UIImage*)image inCellAtIndexPath:(NSIndexPath*)indexPath centeringBounds:(CGRect)normalizedCenterBounds animated:(BOOL)animated ` 
to encapsulate the caching as well and avoid this intertwining telescoping pattern that is here now.